lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp

```diff
@@ -377,7 +377,12 @@ ParsedDWARFTypeAttributes::ParsedDWARFTypeAttributes(const DWARFDIE &die) {
       break;
 
     case DW_AT_object_pointer:
-      object_pointer = form_value.Reference();
+      // GetAttributes follows DW_AT_specification.
+      // DW_TAG_subprogram definitions and declarations may both
+      // have a DW_AT_object_pointer. Don't overwrite the one
+      // we parsed for the definition with the one from the declaration.
```

Comment:
This will only be true if the definition die has `DW_AT_object_pointer` *before* `DW_AT_specification`. `GetAttributes` expands the specification attribute "in place", so if the order is reversed, we will encounter the attribute from the specification first.

I don't know if there's a good reason for the way that `GetAttributes` works. It might be possible to just change it to visit the main DIE first.

Reply:
Good catch. For some reason from an initial glance at GetAttributes I assumed it parsed the specification later. Will see if the iteration order is fixable

Reply:
Not sure if the spec says anything about duplicate attributes across DW_AT_specifications

Reply:
An alternative would be to say that `DW_AT_object_pointer` doesn't apply if we found it through a specification/abstract_origin ([like we do for `DW_AT_sibling` and `DW_AT_declaration` already](https://github.com/llvm/llvm-project/blob/d2e7ee77d33e8b3be3b1d4e9bc5bc4c60b62b554/lldb/source/Plugins/SymbolFile/DWARF/DWARFDebugInfoEntry.cpp#L312-L322)). The object pointer points to a child DIE, so I don't know why we would want to include it in the list of attributes when getting attributes of the definition. That feels like a more consistent/less intrusive approach. Wdyt?

If we ever find that we do for some reason want to include both `DW_AT_obejct_pointer`s in `GetAttributes`, then we can change it then?

Reply:
I can sort of see how that could make sense, though I'm not prepared to call it "more consistent". I'm also unsure how many other attributes could be argued into the list in this way, and I wouldn't want this list to get big.

If you want to go that way, then I think it's fine, though I think I'd prefer to change the iteration order (FWIW, llvm's `DWARFDie::findRecursively` visits the current element first before recursing, although it doesn't have a direct equivalent to our GetAttributes)

Reply:
> FWIW, llvm's DWARFDie::findRecursively visits the current element first before recursing

Cool! I feel like aligning LLDB to do the same is the right thing to do then

Reply:
[This comment](https://github.com/llvm/llvm-project/blob/d2e7ee77d33e8b3be3b1d4e9bc5bc4c60b62b554/lldb/source/Plugins/SymbolFile/DWARF/DWARFDebugInfoEntry.cpp#L292-L293) suggests that users can expect some sort of non-determinism already. So making it stricter doesn't seem harmful

Reply:
Actually that comment sounds like the intention was to only consider an attribute once. Which is not what this function is doing. Wonder if that's something worth trying to do as well. I.e., fix iteration order, but separately also make sure duplicate attributes don't get overwritten? I'm not sure it's ever desirable (examples are `DW_AT_object_pointer` and `DW_AT_low_pc`)

Reply:
Could be, but you also could be reading too much into it. There are other (better) ways to say that the attribute will appear only once and "take *precedence*" could have been meant as "it will *precede* the other attribute in the list" (It's probably not valid English, but I can see someone might say that, and it sort matches what the code does).

That said, I think that omitting the second instance of the attribute should not be a problem functionality-wise, although I am slightly concerned about the performance impact (I guess you'd have to iterate through the list to see if the attribute is not already there). This function is called from the manual dwarf index so it is somewhat performance sensitive. If you want to go down that path, I'd like to measure the impact on indexing speed.

Reply:
The nice thing about omitting the second attribute is that I *think* we could get rid of the `Recurse` argument. Because the only caller that sets `Recurse` to `No` is to avoid possibly overwriting `DW_AT_low_pc` with the one from the declaration. Though haven't thought about this aspect too much yet.

> I am slightly concerned about the performance impact (I guess you'd have to iterate through the list to see if the attribute is not already there)

I was thinking we just make `DWARFAttributes` hold a `SmallSet` instead of a `SmallVector`. And `Append` will only ever insert a single entry (the only caller that `Appends` is currently `GetAttributes`, so if we fix iteration order then making it a set will ensure we only see the attribute of the most specific DIE)

Reply:
> That would also be true with the implementation which puts the most specific attribute first (the code could ignore the second instance, like you do now), although keeping recurse option might still be nice for performance reasons (if we know we don't need to recurse, we can skip doing that).

Makes sense, I'll table the de-duplication for now

Reply:
Gave it a shot in https://github.com/llvm/llvm-project/pull/123261